Add SA1Hybridizer

Atari2 opened the pull request:
This is basically a straight up port of sa1hybridizer (https://github.com/Atari2/SA1Hybridizer) to javascript. 
I also fixed a minor typo in the README.

Atari2 committed:
Add SA1 Hybridizer tool.

Atari2 committed:
Slight renaming and a little improvement.

Atari2 committed:
Fixed a typo in the README

telinc1 commented on src/sa1hybridizer/index.js:
```diff
@@ -0,0 +1,396 @@
+    let bwram_list = [map16_lo_by, map16_hi_by, save_mem]
+    let bwram_indexes = bwram_list.map(x => {
+        let [b, e] = x;
+        return b <= bwram_word <= e;
```
No such syntax exists in JavaScript and this will always return `true`. It's interpreted as `(b <= bwram_word) <= e`. The left-hand side is a boolean, which when cast to a number, will result in e.g. `0 <= e`.

Don't know Python, but I assume this is what you meant?
```suggestion
        return b <= bwram_word && bwram_word <= e;
```

Atari2 replied:
Fixed in 010b017cb417f4e3d9f01b938abf4a2a9697f87c

telinc1 commented on src/sa1hybridizer/index.js:
```diff
@@ -0,0 +1,396 @@
+    }
+    let numeric_word = undefined;
+    try {
+        numeric_word = parseInt(word, 16);
```
`parseInt` doesn't throw errors. For invalid input, it returns `NaN`, which will then spread through the rest of the code (you can detect NaNs with `Number.isNaN`).

Atari2 replied:
Done (and also fixed in another place that did the same thing)

telinc1 commented on src/sa1hybridizer/index.js:
```diff
@@ -0,0 +1,396 @@
+    }
+    let bwram_define_needed = false;
+    if (word.length === 6)
+        word, bwram_define_needed = check_bwram(word);
```
JS interprets this as two separate expressions, `word` (which is essentially a no-op) and `bwram_define_needed = check_bwram(word);`. Since `check_bwram` returns an array, `bwram_define_needed` will always be truthy. I assume you meant to use a destructuring syntax instead.
```suggestion
        [word, bwram_define_needed] = check_bwram(word);
```

Atari2 replied:
Fixed in 010b017cb417f4e3d9f01b938abf4a2a9697f87c

telinc1 commented on src/sa1hybridizer/index.js:
```diff
@@ -0,0 +1,396 @@
+                     a ram address leave it alone\n`);
+                }
+            }
+            outlines[index] = line.trimStart();
```
`trimStart` or `trimEnd`? The Python version uses `rstrip`. This one destroys indentation.

Atari2 replied:
Fixed in 010b017cb417f4e3d9f01b938abf4a2a9697f87c

telinc1 commented on src/sa1hybridizer/index.js:
```diff
@@ -0,0 +1,396 @@
+                        try {
+                            let comma_index;
+                            try {
+                                comma_index = word_tuples[i + 1][0] === WordType.COMMA ? i + 1 : -1;
```
Shouldn't this be `word_tuples[i + 1].type`? That should produce identical behavior to the Python version, though I'm not sure exactly where it expects its `IndexError`s from.

Atari2 replied:
This is fixed now, the IndexError could happen when index in word_tuples with i + 1, however js doesn't throw when accessing out of bounds so I restructured this part.

telinc1 commented on src/sa1hybridizer/index.js:
```diff
@@ -0,0 +1,396 @@
+    });
+    let subs = ['map16_lo_by', 'map16_hi_by', 'save_mem']
+    if (bwram_indexes.some(x => x)) {
+        let true_index = bwram_indexes.indexOf(True)
```
`True` is undefined, `true` is a literal boolean.
```suggestion
        let true_index = bwram_indexes.indexOf(true)
```

Atari2 replied:
Fixed in 010b017cb417f4e3d9f01b938abf4a2a9697f87c

telinc1 reviewed (changes requested):
This port is still too unstable to accept. The code uses (what I assume is) Python syntax that behaves differently in JS and malfunctions as result. For example, I tried it on the [Circling Thwomp](https://smwc.me/s/13893) and it missed a `$7FAB10` that the Python version converted correctly (fwiw, the Python version also modified a `$0020` that should have been left alone, but that's to be expected from an automated tool).

I've pointed out the mistakes that caught my eye, but I've probably missed some. I'm not familiar with Python or with this code, so I don't feel comfortable fixing/testing all of the possible branches myself.

Atari2 committed:
Bugfixes and consistency.

- Ensured consistency by always using `;` at the end of lines.
- Fixed the issues pointed out by @telinc1 as well as a couple of things I noticed while looking throught it.
- Restructured some code to make it clearer what it is doing.

Atari2 committed:
Fixed more instances of begin <= value <= end.

It was a python-ism that didn't translate to JS very well.
Basically translated `begin <= value <= end` to `value >= begin && value <= end` which is the same thing.

Added a function to do it to make it clearer.

I think I replaced all occurrences.

telinc1 reviewed (approved):
Looks good now. Thanks for your contribution!
